TrueCraft.Core.Test/World/SugarCaneDecoratorTests.cs

```diff
@@ -0,0 +1,461 @@
+        public NibbleSlice SkyLight { get; }
+        public IRegion ParentRegion { get; set; }
+        public int GetHeight(byte x, byte z)
+        {
+            // Pretty sure this is always one since we have a flat surface 
+            // (Unless water heights are reported differently)
+            return 1;
+        }
+
+        public void UpdateHeightMap()
+        {
+            throw new NotImplementedException();
+        }
+
+
+        private static byte MapXZToTestBlock(int X, int Z)
+        {
+            // Sand on the outer left hand side
+            if (X == 0)
+            {
+                return GrassBlock.BlockID;
+            }
+            
+            // Sand on the outer right hand side
+            if (X == 5)
+            {
+                return SandBlock.BlockID;
+            }
+
+            if (Z == 0)
+                return WaterBlock.BlockID;
+
+            if (Z == 3)
+                return StationaryWaterBlock.BlockID;
+
+            switch (X)
+            {
+                case 1:
+                    return GrassBlock.BlockID;
+                case 2:
+                    return DirtBlock.BlockID;
+                case 3:
+                    return SandBlock.BlockID;
+                case 4:
+                    return StoneBlock.BlockID;
+            }
+
+            // What's in that block? 
+            // Back at me. 
+            // I have it, it's a block that doesn't matter because it's beyond our test grid.  
+            // Look again
+            // The block is now diamonds. 
+            // Anything is possible when you're in a test mock
+            // I'm on a pig. 
+            return DiamondBlock.BlockID;
+        }
+
+
+        private static Dictionary<Coordinates3D, byte> createStartingBlockDictionary()
+        {
+            int xBounds = 6;
+            int yBounds = 7;
+            int zBounds = 6;
+
+            Dictionary<Coordinates3D, byte> blockDictionary = new Dictionary<Coordinates3D, byte>();
+
+            for (int x = 0; x < xBounds; x++)
+            {
+                for (int z = 0; z < zBounds; z++)
+                {
+                    byte row1Blocks = MapXZToTestBlock(x, z);
+                    for (int y = 0; y < yBounds; y++)
+                    {
+                        byte blockToStore = AirBlock.BlockID;
+                        if (y == 0)
+                        {
+                            // Dirt on the lowest layer
+                            blockToStore = DirtBlock.BlockID;
+                        }
+                        if (y == 1)
+                        {
+                            blockToStore = row1Blocks;
+                        }
+                        blockDictionary.Add(new Coordinates3D(x,y,z), blockToStore);
+                    }
+                }
+            }
+
+            return blockDictionary;
+        }
+
+        private Dictionary<Coordinates3D, byte> blockDictionary = createStartingBlockDictionary();
```

Comment:
You should be able to use the actual block dictionary, no? Saves you a lot of code

Reply:
So I preferred factoring createStartingBlockDictionary() into a method that returned starting state and assign the result to a collection because:

-> I call another method from within that method on each iteration. I found it a bit hard to reason from my hokey mapping of blocks all in one block, so I just split it up. I'd rather have another kind of setup for representing the block layout in the code, which I alluded to in IRC. As is I think the current approach is utter garbage, but it worked and something garbage in test code is a better start than nothing. 
-> It adds a nice seam visually for the inevitable conversion to something reasonably represented. I lift the definition of block dictionary to happen in the constructor, factor out the silly methods to a temporary method that will be used to generate roughly the right representation for tests, plug that representation in instead of using a method, and then call it a day. 
-> I have some bad experiences with code that directly manipulates global collections and found that even a simple method seam made the code easier to reason about for the inevitable fix to use a data file / other representation. 

I can do the work to make this representation for test data better now, if you'd like. I'd rather move on to implementing some other cake related things and then back track to this when if add a simple unit test for spawn uniformity for the CactusDecorator. 

(Actually, I'd rather smoke test the launcher for single player games which would be a high value test, then get more familiar with some of the core systems related to making cake. (Entity spawning + item spawning, block updates + spawning, making sure placing a block from an item can be tested))

Reply:
I don't really follow, probably just because I'm tired. But this test is way too long and you should find a way to improve it. Look at this: https://github.com/SirCmpwn/TrueCraft/blob/master/TrueCraft.Core.Test/Physics/PhysicsEngineTest.cs

Reply:
I'm confused. How is that example useful here? 

```
var world = new TrueCraft.Core.World.World("default", new FlatlandGenerator()); 
```

I don't want a flat land generated here. I want a specific arrangement of blocks which will test most of the specifics of block generation for sugarcane. There are certain invariants that I want to validate and that arrangement of blocks sloppily represented as a method roughly corresponds to that. (Growth on blocks should not occur with blocks diagonally adjacent to water, things don't propagate to other kinds of block, works in both cases where a block can neighbor different kinds of water and grow on both of the kinds of valid blocks) 

Referring to the other aspects of that test: 
I don't want to take the dependency on the IWorld or generators. I just need an IChunk, with blocks in it, which is the interface that the Decorator consumes. 

I think a better way to represent this would be to think about each layer of blocks, and then specify the interesting layers as a clearer array or something in the code, and then you can reason more about what blocks are around. 

Something like: 
~~~ 
[ new AirLayer(), 
/* that repeated four more times */, 
 new InterestingTestPatternLayer(), 
 new DirtLayer() 
] 
~~~ 
Where `interestingTestPatternLayer` has some nicely readable representation defined somewhere. Maybe as an array in the code or in some data file. It may be more readable for it to be in a data file because Chunk.Width, which governs how iteration occurs over the blocks in the IChunk for decorating, is much larger than six, which will make the data more verbose and harder to inline in the source.  This is why I end up having to return diamonds in that method for areas outside of the bounds of my test blocks. Why my IChunk has to have that specific dimension doesn't seem to make sense to me. That's not an invariant implied by its interface or its use cases here. 

There are other tests that would require a test fake for IChunk that could use a specifically specified arrangement of blocks, so it wouldn't make sense for me to remove the test fake that creates an IChunk based on supplied block layouts, and I just do not enjoy maintaining or writing mocks. I don't need to take any of the dependencies of the IWorld beyond its seed. It's not my fault that the interfaces for decorating blocks require me to have implementations that pull in the entire universe. 

And really, they don't have to. I'm reading through the decorators, and excepting two methods that aren't required for this test, the methods implemented in this set of test fakes is all that's required for decorating of any flavor. I've read through most of the decorators and decorations and I'm seeing references JUST to the world seed, and JUST to the relevant IChunk methods I call. I think these tests are trying to tell us that maybe we could use a different interface there. But that's another topic. 

The test fakes here don't have to be defined in this file, and really their implementation details don't matter unless you break the test and evaluate that you need to review the block layout. So the test file itself condenses down to the 10 lines of code for the test method (plus the other test method you wanted me to add 😸 ) I'll extract the fakes to separate files so the test method class is much simpler to read.

Reply:
I'm talking about how that test handles the block repository.

Reply:
You can add a mock for IWorld and only mock its seed, you know?

Reply:
tldr: 
I want to fix this, but like, come on man there's bigger fish to fry. (I've tried two different windows 10 installs and I cannot get the client to start via the launcher or otherwise, and I haven't figured out where to find the logs to help me debug the issue yet. (Low prio to me because I can unit test independent of rendering)) And there's at least two crashing bugs related to singleplayer (relevant to my interests) reported that could/should be validated. #275 #276 I'd rather work on those now that I've tasted the code. 

I'll have to come back to this kind of test when I write code to validate wheat growing mechanics (important for making cake) anyways, so like, for now I just want to say YAGNI on gorgeous testing here and move onto more important things in the interim. 

So I propose: 
Extracting test fakes into seperate files so they don't clutter this file
Implementing the extra test methods you suggested below. 
Handling cleaning up the representation of the block layout later.

If it's a problem and you just don't want to merge this, that's fine too. I can pick this up later or just drop the unit test Kappa.

Reply:
Do you understand how mocks work?

Reply:
Why mock? If I create an instance of a mock to pass around to multiple tests, couldn't I have just implemented the interface with all I needed for the test? 

In this case especially, since the block collection needs to take updates, the mock would basically just be an implementation of the class, but we'd hide the interface methods we don't need to update.

Reply:
+ It's fine if you don't like this test fake code, but seriously do the interfaces consumed at these methods need to be so monstrous? Do you really need an IWorld for the way that the decoration code is written?